# DeepSeek-Coder-7B-Instruct v1.5 Evaluation on LeetCode Dataset

This notebook evaluates DeepSeek-Coder-7B-Instruct v1.5 on the LeetCode contests dataset using **Execution Prediction** and related reasoning benchmarks.

1. **Introduction of Deepseek-Coder-7B-Instruct v1.5**

Deepseek-Coder-7B-Instruct-v1.5 is continue pre-trained from Deepseek-LLM 7B on 2T tokens by employing a window size of 4K and next token prediction objective, and then fine-tuned on 2B tokens of instruction data.

Home Page: [DeepSeek](https://www.deepseek.com)  \
Repository: [deepseek-ai/deepseek-coder](https://huggingface.co/deepseek-ai/deepseek-coder)  \
Chat With DeepSeek Coder: [DeepSeek-Coder](https://chat.deepseek.com/coder)

2. **Evaluation Results**

DeepSeek Coder

3. **How to Use**

Here give some examples of how to use our model.

*Chat Model Inference*

```python
from transformers import AutoTokenizer, AutoModelForCausalLM
tokenizer = AutoTokenizer.from_pretrained("deepseek-ai/deepseek-coder-7b-instruct-v1.5", trust_remote_code=True)
model = AutoModelForCausalLM.from_pretrained("deepseek-ai/deepseek-coder-7b-instruct-v1.5", trust_remote_code=True).cuda()
messages=[
    { 'role': 'user', 'content': "write a quick sort algorithm in python."}
]
inputs = tokenizer.apply_chat_template(messages, add_generation_prompt=True, return_tensors="pt").to(model.device)

outputs = model.generate(inputs, max_new_tokens=512, do_sample=False, top_k=50, top_p=0.95, num_return_sequences=1, eos_token_id=tokenizer.eos_token_id)
print(tokenizer.decode(outputs[0][len(inputs[0]):], skip_special_tokens=True))
```

4. **License**

This code repository is licensed under the MIT License. The use of DeepSeek Coder models is subject to the Model License. DeepSeek Coder supports commercial use.

See the LICENSE-MODEL for more details.

5. **Contact**

If you have any questions, please raise an issue or contact us at service@deepseek.com.



## Step 1: Setup Environment

Install required packages for mxfp4 quantization support.

In [1]:
# Install required libraries
!pip install -q --upgrade pip
!pip install -q transformers accelerate bitsandbytes datasets


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 899.7/899.7 MB 1.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 594.3/594.3 MB 2.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.2/10.2 MB 150.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 88.0/88.0 MB 30.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 954.8/954.8 kB 61.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 193.1/193.1 MB 13.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.2/1.2 MB 65.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 63.6/63.6 MB 14.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 267.5/267.5 MB 4.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 288.2/288.2 MB 4.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 322.3/322.3 MB 3.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 39.3/39.3 MB 55.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 

⚠️ **IMPORTANT**: Please restart your Colab runtime after running the cell above.

Click: **Runtime → Restart runtime**

## Step 2: Load DeepSeek-Coder-7B-Instruct v1.5 Model

In [2]:
from transformers import AutoModelForCausalLM, AutoTokenizer
import torch

print("Loading DeepSeek-Coder-7B-Instruct v1.5 model...")
print(f"CUDA available: {torch.cuda.is_available()}")
print(f"CUDA device: {torch.cuda.get_device_name(0) if torch.cuda.is_available() else 'None'}")

MODEL_ID = "deepseek-ai/deepseek-coder-7b-instruct-v1.5"

tokenizer = AutoTokenizer.from_pretrained(
    MODEL_ID,
    trust_remote_code=True,
    use_fast=True,
)

model = AutoModelForCausalLM.from_pretrained(
    MODEL_ID,
    device_map="auto",
    trust_remote_code=True,
    load_in_4bit=True,
    bnb_4bit_compute_dtype=torch.float16,
)

print("✓ Model loaded successfully!")


Loading gpt-oss-20b model...
CUDA available: True
CUDA device: NVIDIA A100-SXM4-40GB


/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


tokenizer_config.json: 0.00B [00:00, ?B/s]

tokenizer.json:   0%|          | 0.00/27.9M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/98.0 [00:00<?, ?B/s]

chat_template.jinja: 0.00B [00:00, ?B/s]

config.json: 0.00B [00:00, ?B/s]

`torch_dtype` is deprecated! Use `dtype` instead!


model.safetensors.index.json: 0.00B [00:00, ?B/s]

Fetching 3 files:   0%|          | 0/3 [00:00<?, ?it/s]

model-00001-of-00002.safetensors:   0%|          | 0.00/4.80G [00:00<?, ?B/s]

model-00000-of-00002.safetensors:   0%|          | 0.00/4.79G [00:00<?, ?B/s]

model-00002-of-00002.safetensors:   0%|          | 0.00/4.17G [00:00<?, ?B/s]

Fetching 41 files:   0%|          | 0/41 [00:00<?, ?it/s]

__init__.cpython-312.pyc:   0%|          | 0.00/220 [00:00<?, ?B/s]

_ops.py:   0%|          | 0.00/201 [00:00<?, ?B/s]

__init__.py:   0%|          | 0.00/363 [00:00<?, ?B/s]

_common.py: 0.00B [00:00, ?B/s]

matmul_ogs.py: 0.00B [00:00, ?B/s]

compaction.py: 0.00B [00:00, ?B/s]

_masked_compaction.py:   0%|          | 0.00/814 [00:00<?, ?B/s]

_finalize_matmul.py: 0.00B [00:00, ?B/s]

_p_matmul_ogs.py: 0.00B [00:00, ?B/s]

_matmul_ogs.py: 0.00B [00:00, ?B/s]

opt_flags.py: 0.00B [00:00, ?B/s]

opt_flags_intel.py: 0.00B [00:00, ?B/s]

numerics.py: 0.00B [00:00, ?B/s]

opt_flags_nvidia.py: 0.00B [00:00, ?B/s]

opt_flags_amd.py: 0.00B [00:00, ?B/s]

_downcast_to_mxfp.py: 0.00B [00:00, ?B/s]

flexpoint.py: 0.00B [00:00, ?B/s]

mxfp.py: 0.00B [00:00, ?B/s]

_upcast_from_mxfp.py: 0.00B [00:00, ?B/s]

routing.py: 0.00B [00:00, ?B/s]

proton_opts.py:   0%|          | 0.00/456 [00:00<?, ?B/s]

reduce_bitmatrix.py: 0.00B [00:00, ?B/s]

_expt_data.py: 0.00B [00:00, ?B/s]

_routing_compute.py: 0.00B [00:00, ?B/s]

swiglu.py: 0.00B [00:00, ?B/s]

_swiglu.py: 0.00B [00:00, ?B/s]

specialize.py: 0.00B [00:00, ?B/s]

layout.py: 0.00B [00:00, ?B/s]

tensor.py: 0.00B [00:00, ?B/s]

target_info.py: 0.00B [00:00, ?B/s]

base.py:   0%|          | 0.00/352 [00:00<?, ?B/s]

strided.py:   0%|          | 0.00/337 [00:00<?, ?B/s]

hopper_value.py: 0.00B [00:00, ?B/s]

blackwell_scale.py: 0.00B [00:00, ?B/s]

hopper_scale.py: 0.00B [00:00, ?B/s]

_topk_backward.py: 0.00B [00:00, ?B/s]

testing.py: 0.00B [00:00, ?B/s]

topk.py: 0.00B [00:00, ?B/s]

_topk_forward.py: 0.00B [00:00, ?B/s]

Loading checkpoint shards:   0%|          | 0/3 [00:00<?, ?it/s]

Fetching 41 files:   0%|          | 0/41 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/177 [00:00<?, ?B/s]

✓ Model loaded successfully!


## Step 3: Load LeetCode Dataset from HuggingFace

In [ ]:
from datasets import load_dataset

DATASET_REPO_ID = "asgaur/leetcode-contests-431-467-mutations-strat110"
SPLIT = "train"

print(f"Loading dataset from {DATASET_REPO_ID}...")
dataset = load_dataset(DATASET_REPO_ID)

print(f"\n✓ Dataset loaded!")
print(f"Total samples: {len(dataset[SPLIT])}")
sample = dataset[SPLIT][0]
print(f"First sample: {sample['id']} ({sample['difficulty']})")


Loading dataset from asgaur/leetcode-contests-431-467-mutated...


README.md: 0.00B [00:00, ?B/s]

data/train-00000-of-00001.parquet:   0%|          | 0.00/226k [00:00<?, ?B/s]

Generating train split:   0%|          | 0/347 [00:00<?, ? examples/s]


✓ Dataset loaded!
Total samples: 347

First sample:
  ID: contest431_q3702_s0
  Function: maxLength
  Difficulty: easy
  Input: maxLength(nums=[1,2,1,2,1,1,1])...


## Step 4: Define Helper Functions

In [14]:
import json
import re
import ast
from typing import Dict, Tuple, Optional

def build_execution_prediction_prompt(sample: Dict, use_mutated: bool = False) -> str:
    """
    Build execution prediction prompt for either the original or mutated program.
    """
    function_name = sample['function_name']
    code_key = 'mutated_code' if use_mutated else 'code'
    program = sample.get(code_key)
    test_input = sample['input']

    if test_input and test_input.startswith(f"{function_name}(") and test_input.endswith(")"):
        input_args = test_input[len(function_name) + 1:-1]
    else:
        input_args = test_input

    prompt = f"""You are given a Python program and an assertion containing an input to a function. Replace the ?? in the assertion with a literal (no unsimplified expressions, no function calls) representing the function's return value for the given input. Execute the program exactly as written, even if it is incorrect or incomplete. For your final answer, provide the full assertion in [ANSWER] and [/ANSWER] tags.

[PYTHON]
{program}
assert {function_name}({input_args}) == ??
[/PYTHON]"""

    return prompt


def build_execution_choice_prompt(sample: Dict, original_first: bool = True) -> Tuple[str, Dict[str, str]]:
      """Build the execution-choice prompt and map A/B to original/mutated."""
      function_name = sample['function_name']
      original_code = sample['code']
      mutated_code = sample.get('mutated_code') or original_code
      test_input = sample['input']

      if test_input and test_input.startswith(f"{function_name}(") and test_input.endswith(")"):
          input_args = test_input[len(function_name) + 1:-1]
      else:
          input_args = test_input

      if original_first:
          program_a, program_b = original_code, mutated_code
          mapping = {'A': 'original', 'B': 'mutated'}
      else:
          program_a, program_b = mutated_code, original_code
          mapping = {'A': 'mutated', 'B': 'original'}

      prompt = (
          "You are given two Python programs below and an assertion containing an input to a function. "
          "First, choose either program, whichever one you are more confident in reasoning about. "
          "Then, replace the ?? in the assertion with a literal (no unsimplified expressions, no function calls) "
          "representing the function's return value for the given input on your chosen program. Execute the program "
          "exactly as written, even if it is incorrect or incomplete. For your final answer, output the letter of your "
          "chosen program (A or B) and the full assertion in the following json format:\n\n"
          "{{\n"
          '  "chosen_program": "A or B",\n'
          '  "assertion": "full_assertion"\n'
          "}}\n\n"
          f"[PROGRAM_A]\n{program_a}\n[/PROGRAM_A]\n"
          f"[PROGRAM_B]\n{program_b}\n[/PROGRAM_B]\n"
          f"[ASSERTION]\nassert {function_name}({input_args}) == ??\n[/ASSERTION]"
      )

      return prompt, mapping


def parse_execution_choice_response(response: str) -> Dict[str, str]:
    """
    Parse the model response for execution choice into a structured dictionary.
    Expects a JSON blob containing "chosen_program" and "assertion" keys.
    """
    json_match = re.search(r'\{\s*"chosen_program"\s*:.*?\}', response, re.DOTALL)
    if not json_match:
        raise ValueError('Could not find JSON payload in the response.')

    json_text = json_match.group(0)
    try:
        return json.loads(json_text)
    except json.JSONDecodeError:
        chosen_match = re.search(r'"chosen_program"\s*:\s*"?([A-Za-z])"?', json_text)
        assertion_match = re.search(r'"assertion"\s*:\s*("(?:[^"\]|\.)*")', json_text)
        if not chosen_match or not assertion_match:
            raise ValueError('Failed to parse execution choice JSON response.')

        chosen_program = chosen_match.group(1)
        assertion_literal = assertion_match.group(1)
        try:
            assertion = ast.literal_eval(assertion_literal)
        except Exception:
            assertion = assertion_literal.strip('"')

        return {
            'chosen_program': chosen_program,
            'assertion': assertion,
        }


def extract_output_from_assertion(assertion: str) -> str:
    """Extract the predicted output value from an assertion string."""
    if not assertion:
        return ''

    text = assertion.strip()
    text = re.sub(r'^\[ASSERTION\]\s*', '', text, flags=re.IGNORECASE)
    text = re.sub(r'\s*\[/ASSERTION\]$', '', text, flags=re.IGNORECASE)

    match = re.search(r'assert\s+[\w\.]+\([^)]*\)\s*==\s*(.+)', text)
    if match:
        return match.group(1).strip()

    return text


def extract_answer_from_response(response: str) -> str:
    """
    Extract predicted answer from [ANSWER] tags.

    Expected format: [ANSWER] assert function_name(input) == output [/ANSWER]
    """
    pattern = r'\[ANSWER\](.*?)\[/ANSWER\]'
    matches = re.findall(pattern, response, re.DOTALL | re.IGNORECASE)
    if matches:
        assertion = matches[0].strip()
        match = re.search(r'assert\s+\w+\([^)]*\)\s*==\s*(.+)', assertion)
        if match:
            return match.group(1).strip()
        return assertion

    pattern = r'assert\s+\w+\([^)]*\)\s*==\s*(.+?)(?:|$)'
    matches = re.findall(pattern, response, re.MULTILINE)
    if matches:
        return matches[0].strip()

    return response.strip()


def check_predicted_output(predicted_output: str, expected_output: str) -> Tuple[bool, str]:
    """
    Compare predicted output with expected output.
    """
    try:
        predicted = (predicted_output or '').strip()
        expected = (expected_output or '').strip()

        if predicted == expected:
            return True, None

        try:
            predicted_val = ast.literal_eval(predicted)
            expected_val = ast.literal_eval(expected)
            if predicted_val == expected_val:
                return True, None
        except (ValueError, SyntaxError):
            pass

        return False, f"Predicted: {predicted}, Expected: {expected}"
    except Exception as exc:
        return False, str(exc)


def is_boolean_output(value: str) -> bool:
    """
    Determine whether a serialized output represents a Boolean value.
    """
    if value is None:
        return False

    try:
        parsed = ast.literal_eval(value.strip())
        return isinstance(parsed, bool)
    except (ValueError, SyntaxError, AttributeError):
        lowered = value.strip().lower()
        return lowered in {'true', 'false'}


print('✓ Helper functions defined (execution prediction + metrics)')



✓ Helper functions defined (execution prediction + metrics)


## Step 5: Test Execution Prediction on One Sample

This cell demonstrates the **Execution Prediction** task:
1. Model receives a Python program and an assertion with `??`
2. Model predicts what the output will be
3. Model provides answer in `[ANSWER]...[/ANSWER]` tags

In [5]:
# Test with one sample
test_sample = dataset['train'][0]

prompt = build_execution_prediction_prompt(test_sample)
print("Prompt:")
print(prompt)
print("" + "="*60)

# Generate prediction with gpt-oss
messages = [
    {"role": "user", "content": prompt}
]

inputs = tokenizer.apply_chat_template(
    messages,
    add_generation_prompt=True,
    return_tensors="pt",
    return_dict=True,
    reasoning_effort="medium",  # Can be "low", "medium", or "high"
).to(model.device)

print("Generating prediction...")
# Increase max_new_tokens to allow for reasoning + answer
generated = model.generate(**inputs, max_new_tokens=1000)
response = tokenizer.decode(generated[0][inputs["input_ids"].shape[-1]:], skip_special_tokens=True)

print("
Generated Response:")
print(response)
print("" + "="*60)

# Extract predicted output
predicted_output = extract_answer_from_response(response)
print("Extracted Predicted Output:")
print(predicted_output)
print("" + "="*60)

# Check correctness
is_correct, error = check_predicted_output(
    predicted_output,
    test_sample['output']
)

print(f"
Test Result: {'✓ CORRECT' if is_correct else '✗ INCORRECT'}")
if error:
    print(f"Error: {error}")
print(f"Expected: {test_sample['output']}")


✓ Helper functions defined (execution prediction + metrics)


## Step 6: Benchmark Execution Prediction (OC / OR / MC / MR)

Generate multiple samples per problem to estimate pass@1 for original/mutated correctness and reversion metrics.

In [6]:

import random
import time
import torch
from typing import Dict, List

import pandas as pd
from tqdm.auto import tqdm

NUM_PROBLEMS = None  # Evaluate the full stratified split by default
START_INDEX = 0
NUM_GENERATIONS = 5  # Paper uses five generations per problem for pass@1
MAX_NEW_TOKENS = 800
TEMPERATURE = 0.2  # As recommended in the paper for non-reasoning models
TOP_P = 0.95
SEED = 42
SKIP_BOOLEAN_FOR_REVERSION = True  # Exclude booleans from OR/MR, per paper

REASONING_EFFORT = "low"  # Placeholder flag for consistency with Step 5

torch.manual_seed(SEED)
random.seed(SEED)
device = next(model.parameters()).device
pad_token_id = tokenizer.eos_token_id if tokenizer.eos_token_id is not None else tokenizer.pad_token_id


def _generate_response(prompt: str, seed: int) -> tuple[str, float]:
    torch.manual_seed(seed)
    if torch.cuda.is_available():
        torch.cuda.manual_seed_all(seed)
    encoded = tokenizer.apply_chat_template(
        [{"role": "user", "content": prompt}],
        add_generation_prompt=True,
        return_tensors="pt",
        return_dict=True,
    )
    encoded = {k: v.to(device) for k, v in encoded.items()}
    start_time = time.time()
    outputs = model.generate(
        input_ids=encoded["input_ids"],
        attention_mask=encoded.get("attention_mask"),
        max_new_tokens=MAX_NEW_TOKENS,
        temperature=TEMPERATURE,
        top_p=TOP_P,
        do_sample=True,
        pad_token_id=pad_token_id,
        eos_token_id=tokenizer.eos_token_id,
    )
    latency = time.time() - start_time
    generated_tokens = outputs[0, encoded["input_ids"].shape[-1]:]
    response_text = tokenizer.decode(generated_tokens, skip_special_tokens=True)
    return response_text, latency


def _compute_pass(counts: Dict[str, int]) -> float | None:
    return counts["success"] / counts["total"] if counts["total"] else None

split = dataset[SPLIT]
stop = len(split) if NUM_PROBLEMS is None else min(len(split), START_INDEX + NUM_PROBLEMS)
problem_indices = list(range(START_INDEX, stop))
if not problem_indices:
    raise ValueError("No problems selected. Adjust START_INDEX/NUM_PROBLEMS.")

metrics_counts = {
    "OC": {"success": 0, "total": 0},
    "OR": {"success": 0, "total": 0},
    "MC": {"success": 0, "total": 0},
    "MR": {"success": 0, "total": 0},
}
reversion_skip_count = 0
prediction_records: List[Dict] = []
all_latencies: List[float] = []

for idx in tqdm(problem_indices, desc="Execution prediction"):
    sample = split[idx]
    mutated_code = sample.get("mutated_code")
    if not mutated_code:
        prediction_records.append(
            {
                "problem_index": int(idx),
                "problem_id": sample.get("id"),
                "skipped": True,
                "skip_reason": "missing mutated_code",
            }
        )
        continue

    original_output = sample["output"]
    mutated_output = sample.get("mutated_output") or original_output
    include_reversion = True
    if SKIP_BOOLEAN_FOR_REVERSION and (
        is_boolean_output(original_output) or is_boolean_output(mutated_output)
    ):
        include_reversion = False
        reversion_skip_count += 1

    original_prompt = build_execution_prediction_prompt(sample, use_mutated=False)
    mutated_prompt = build_execution_prediction_prompt(sample, use_mutated=True)

    oc_successes = or_successes = mc_successes = mr_successes = 0
    original_predictions = []
    mutated_predictions = []

    seed_base = SEED + idx * 1000
    for gen_idx in range(NUM_GENERATIONS):
        response_text, latency = _generate_response(original_prompt, seed_base + gen_idx)
        all_latencies.append(latency)
        original_prediction = extract_answer_from_response(response_text)
        oc_correct, _ = check_predicted_output(original_prediction, original_output)
        or_correct = None
        if include_reversion:
            or_correct, _ = check_predicted_output(original_prediction, mutated_output)
        original_predictions.append(
            {
                "generation": gen_idx,
                "prediction": original_prediction,
                "response": response_text,
                "latency_s": latency,
                "oc_correct": bool(oc_correct),
                "or_correct": bool(or_correct) if isinstance(or_correct, bool) else None,
            }
        )
        if oc_correct:
            oc_successes += 1
        if include_reversion and or_correct:
            or_successes += 1

        response_text_mut, latency_mut = _generate_response(mutated_prompt, seed_base + 500 + gen_idx)
        all_latencies.append(latency_mut)
        mutated_prediction = extract_answer_from_response(response_text_mut)
        mc_correct, _ = check_predicted_output(mutated_prediction, mutated_output)
        mr_correct = None
        if include_reversion:
            mr_correct, _ = check_predicted_output(mutated_prediction, original_output)
        mutated_predictions.append(
            {
                "generation": gen_idx,
                "prediction": mutated_prediction,
                "response": response_text_mut,
                "latency_s": latency_mut,
                "mc_correct": bool(mc_correct),
                "mr_correct": bool(mr_correct) if isinstance(mr_correct, bool) else None,
            }
        )
        if mc_correct:
            mc_successes += 1
        if include_reversion and mr_correct:
            mr_successes += 1

    metrics_counts["OC"]["success"] += oc_successes
    metrics_counts["OC"]["total"] += NUM_GENERATIONS
    metrics_counts["MC"]["success"] += mc_successes
    metrics_counts["MC"]["total"] += NUM_GENERATIONS
    if include_reversion:
        metrics_counts["OR"]["success"] += or_successes
        metrics_counts["OR"]["total"] += NUM_GENERATIONS
        metrics_counts["MR"]["success"] += mr_successes
        metrics_counts["MR"]["total"] += NUM_GENERATIONS

    prediction_records.append(
        {
            "problem_index": int(idx),
            "problem_id": sample.get("id"),
            "difficulty": sample.get("difficulty"),
            "include_reversion": include_reversion,
            "original_output": original_output,
            "mutated_output": mutated_output,
            "oc_successes": oc_successes,
            "or_successes": or_successes if include_reversion else None,
            "mc_successes": mc_successes,
            "mr_successes": mr_successes if include_reversion else None,
            "original_predictions": original_predictions,
            "mutated_predictions": mutated_predictions,
        }
    )

metrics_summary = {metric: _compute_pass(counts) for metric, counts in metrics_counts.items()}
avg_latency = (sum(all_latencies) / len(all_latencies)) if all_latencies else None
benchmark_summary = {
    "dataset": DATASET_REPO_ID,
    "problems_evaluated": len(problem_indices),
    "generations_per_problem": NUM_GENERATIONS,
    "oc_pass_at_1": metrics_summary["OC"],
    "or_pass_at_1": metrics_summary["OR"],
    "mc_pass_at_1": metrics_summary["MC"],
    "mr_pass_at_1": metrics_summary["MR"],
    "avg_latency_s": avg_latency,
    "reversion_skipped_problems": reversion_skip_count if SKIP_BOOLEAN_FOR_REVERSION else 0,
}

benchmark_table = pd.DataFrame(
    [
        {"Metric": "Dataset", "Value": benchmark_summary["dataset"]},
        {"Metric": "Problems Evaluated", "Value": benchmark_summary["problems_evaluated"]},
        {"Metric": "Generations per Problem", "Value": benchmark_summary["generations_per_problem"]},
        {"Metric": "OC pass@1", "Value": metrics_summary["OC"]},
        {"Metric": "OR pass@1", "Value": metrics_summary["OR"]},
        {"Metric": "MC pass@1", "Value": metrics_summary["MC"]},
        {"Metric": "MR pass@1", "Value": metrics_summary["MR"]},
        {"Metric": "Avg latency (s)", "Value": benchmark_summary["avg_latency_s"]},
        {"Metric": "Reversion skipped", "Value": benchmark_summary["reversion_skipped_problems"]},
    ]
)
formatters = {
    "Value": (lambda val: f"{val:.3f}" if isinstance(val, float) and val is not None else val)
}

print("✓ Execution prediction benchmark complete!")



Evaluating 20 LeetCode problems with 5 generations each.
Reasoning effort: low, temperature: 0.6, top_p: 0.95


Evaluating problems:   0%|          | 0/20 [00:00<?, ?it/s]

Streaming output truncated to the last 5000 lines.
Coins:0+1+2+(-3)+4=4

Path DRDR: 0->(0,1)->(1,1)->(1,2)->(2,2)
0+1+(-2)+3+4=6

Path DRRD: 0->(0,1)->(0,2)->(1,2)->(2,2)
0+1+(-1)+3+4=7

Path RDDR: 0->(0,1)->(1,1)->(2,1)->(2,2)
0+1+(-2)+(-3)+4=0

Path RDRD: 0->(0,1)->(1,1)->(1,2)->(2,2) same as DRDR? Wait RDRD path: 0,0 -> (0,1) -> (1,1) -> (1,2) -> (2,2) same as DRDR. So 6.

Path RRDD: 0->(0,1)->(0,2)->(1,2)->(2,2) same as DRRD? Actually RRDD: 0,0->(0,1)->(0,2)->(1,2)->(2,2) same as DRRD. So duplicates.

Thus max base is 7.

Now with magic up to 2. Magic can replace coin with max(coin,0). So negative coins become 0. Also can choose to use magic at some step to skip negative.

Let's try path DRRD: coins: 0,1,-1,3,4. Negative at (0,2). Use magic there: replace -1 with 0. Sum becomes 0+1+0+3+4=8. Magic used 1.

Could also use magic at other negative? none. So 8.

Path RDDR: coins 0,1,-2,-3,4. Negatives at (1,1) and (2,1). Use magic twice: replace both with 0. Sum 0+1+0+0+4=5.

Path DRDR:

Counts (success / total generations):
  OC: 53 / 100
  OR: 6 / 100
  MC: 45 / 100
  MR: 2 / 100
Average latency per generation: 46.79s
Problems skipped for reversion metrics (boolean outputs): 0


## Step 7: Save Results

In [ ]:

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd

if 'metrics_summary' not in globals():
    raise RuntimeError('Run Step 6 to compute execution-prediction metrics before plotting.')

metrics = {
    'OC': metrics_summary.get('OC'),
    'OR': metrics_summary.get('OR'),
    'MC': metrics_summary.get('MC'),
    'MR': metrics_summary.get('MR'),
}
labels = list(metrics.keys())
values = [metrics[k] * 100 if metrics[k] is not None else None for k in labels]

plt.figure(figsize=(8, 4))
bars = plt.bar(labels, [v if v is not None else 0 for v in values], color=['#1f77b4', '#ff7f0e', '#2ca02c', '#d62728'])
for bar, val in zip(bars, values):
    if val is None:
        bar.set_alpha(0.3)
        plt.text(bar.get_x() + bar.get_width() / 2, bar.get_height(), 'N/A', ha='center', va='bottom')
    else:
        plt.text(bar.get_x() + bar.get_width() / 2, bar.get_height(), f'{val:.1f}%', ha='center', va='bottom')
plt.ylim(0, 100)
plt.ylabel('pass@1 (%)')
plt.title('Execution Prediction Metrics (OC/OR/MC/MR)')
plt.show()
plt.close()

if 'benchmark_table' not in globals():
    benchmark_table = pd.DataFrame(
        [
            {'Metric': k, 'Value': v}
            for k, v in benchmark_summary.items()
        ]
    )
    formatters = {
        'Value': (lambda val: f"{val:.3f}" if isinstance(val, float) and val is not None else val)
    }

display(benchmark_table.style.format(formatters))



In [7]:
import json
import math
from datetime import datetime

if 'benchmark_summary' not in globals():
    raise RuntimeError('Run the benchmark cell first to produce metrics.')

timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
output_filename = f"deepseek_coder_7b_instruct_v1_5_leetcode_benchmark_{timestamp}.json"

def _clean_nan(value):
    if value is None:
        return None
    if isinstance(value, float) and math.isnan(value):
        return None
    return value

payload = {
    "model": "deepseek-coder-7b-instruct-v1.5",
    "dataset": benchmark_summary["dataset"],
    "reasoning_effort": REASONING_EFFORT,
    "num_problems": benchmark_summary["problems_evaluated"],
    "num_generations": benchmark_summary["generations_per_problem"],
    "temperature": TEMPERATURE,
    "top_p": TOP_P,
    "seed": SEED,
    "skip_boolean_for_reversion": SKIP_BOOLEAN_FOR_REVERSION,
    "reversion_skipped_problems": benchmark_summary["reversion_skipped_problems"],
    "metrics": {k: _clean_nan(v) for k, v in metrics_summary.items()},
    "metrics_counts": metrics_counts,
    "benchmark_summary": {k: _clean_nan(v) for k, v in benchmark_summary.items()},
    "results": results,
}

if 'execution_choice_summary' in globals():
    payload['execution_choice_summary'] = {k: _clean_nan(v) for k, v in execution_choice_summary.items()}
    payload['execution_choice_counts'] = execution_choice_counts
    payload['execution_choice_results'] = execution_choice_results
    payload['execution_choice_config'] = {
        'num_problems': NUM_PROBLEMS_CHOICE,
        'start_index': START_INDEX_CHOICE,
        'runs_per_problem': NUM_RUNS_PER_PROBLEM,
        'reasoning_effort': REASONING_EFFORT_CHOICE,
        'max_new_tokens': MAX_NEW_TOKENS_CHOICE,
        'temperature': TEMPERATURE_CHOICE,
        'top_p': TOP_P_CHOICE,
        'seed': SEED_CHOICE,
        'skip_boolean_for_reversion': SKIP_BOOLEAN_FOR_REVERSION_CHOICE,
    }

with open(output_filename, "w") as f:
    json.dump(payload, f, indent=2)

print(f"✓ Saved evaluation summary to {output_filename}")

try:
    from google.colab import files
    files.download(output_filename)
except ImportError:
    print("(Optional) Run inside Colab to download the file automatically.")

# Execution choice configuration
NUM_PROBLEMS_CHOICE = None
START_INDEX_CHOICE = 0
NUM_RUNS_PER_PROBLEM = 2
REASONING_EFFORT_CHOICE = "low"
MAX_NEW_TOKENS_CHOICE = 800
TEMPERATURE_CHOICE = 0.6
TOP_P_CHOICE = 0.95
SEED_CHOICE = 123
SKIP_BOOLEAN_FOR_REVERSION_CHOICE = True

torch.manual_seed(SEED_CHOICE)
random.seed(SEED_CHOICE)


✓ Saved evaluation summary to gpt_oss_20b_leetcode_benchmark_20251024_172049.json


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

## Step 8: Execution Choice Benchmark (Preference / Correctness / Reversion)

Run the paired-program experiment to measure program preference, correctness, and reversion with order swapping.

In [ ]:
# Plot execution choice metrics
import matplotlib.pyplot as plt
import numpy as np

choice_metrics = {
    'Preference (Original)': execution_choice_summary.get('preference_original'),
    'OC Correct': execution_choice_summary.get('oc_correct'),
    'OR Reversion': execution_choice_summary.get('or_reversion'),
    'MC Correct': execution_choice_summary.get('mc_correct'),
    'MR Reversion': execution_choice_summary.get('mr_reversion'),
}
labels_choice = list(choice_metrics.keys())
values_choice = [choice_metrics[k] * 100 if choice_metrics[k] is not None else None for k in labels_choice]

plt.figure(figsize=(9, 4))
bars = plt.bar(labels_choice, [v if v is not None else 0 for v in values_choice], color=['#1f77b4', '#ff7f0e', '#2ca02c', '#9467bd', '#d62728'])
plt.xticks(rotation=20, ha='right')
for bar, val in zip(bars, values_choice):
    if val is None:
        bar.set_alpha(0.3)
        plt.text(bar.get_x() + bar.get_width() / 2, bar.get_height(), 'N/A', ha='center', va='bottom')
    else:
        plt.text(bar.get_x() + bar.get_width() / 2, bar.get_height(), f'{val:.1f}%', ha='center', va='bottom')
plt.ylim(0, 100)
plt.ylabel('Percentage (%)')
plt.title('Execution Choice Metrics (Preference / Correctness / Reversion)')
plt.tight_layout()
plt.show()
plt.close()

display(execution_choice_table.style.format(choice_formatters))

print('Counts (success / total runs):')
for metric in ['OC', 'OR', 'MC', 'MR']:
    success, total = choice_metric_counts.get(metric, (0, 0))
    value = choice_metric_pass.get(metric)
    percent = f"{value*100:.1f}%" if value is not None else 'N/A'
    print(f"  {metric}: {success} / {total} ({percent})")


In [17]:
import random
import time
from typing import Optional
import pandas as pd

from tqdm.auto import tqdm

# Execution choice configuration
NUM_PROBLEMS_CHOICE = 20          # Set to None to evaluate the full dataset
START_INDEX_CHOICE = 0
NUM_RUNS_PER_PROBLEM = 2          # Two runs per problem with swapped ordering
REASONING_EFFORT_CHOICE = "low"
MAX_NEW_TOKENS_CHOICE = 5000
TEMPERATURE_CHOICE = 0.6
TOP_P_CHOICE = 0.95
SEED_CHOICE = 123
SKIP_BOOLEAN_FOR_REVERSION_CHOICE = True

if NUM_RUNS_PER_PROBLEM not in (1, 2):
    raise ValueError('NUM_RUNS_PER_PROBLEM must be 1 or 2 for the ordering swap protocol.')

torch.manual_seed(SEED_CHOICE)
random.seed(SEED_CHOICE)


def generate_choice_response(prompt: str, seed: int) -> dict:
    """Generate a model response for the execution choice prompt."""
    generator = torch.Generator(device=model.device)
    generator.manual_seed(seed)
    messages = [{"role": "user", "content": prompt}]
    inputs = tokenizer.apply_chat_template(
        messages,
        add_generation_prompt=True,
        return_tensors="pt",
        return_dict=True,
        reasoning_effort=REASONING_EFFORT_CHOICE,
    ).to(model.device)

    start_time = time.time()
    generation_kwargs = {
        "max_new_tokens": MAX_NEW_TOKENS_CHOICE,
        "do_sample": True,
        "temperature": TEMPERATURE_CHOICE,
        "top_p": TOP_P_CHOICE
    }

    pad_id = tokenizer.eos_token_id if tokenizer.eos_token_id is not None else tokenizer.pad_token_id
    if pad_id is not None:
        generation_kwargs["pad_token_id"] = pad_id

    output_ids = model.generate(**inputs, **generation_kwargs)
    latency = time.time() - start_time

    response = tokenizer.decode(
        output_ids[0][inputs["input_ids"].shape[-1]:],
        skip_special_tokens=True,
    )

    return {
        "response": response,
        "latency_s": latency,
    }


if NUM_PROBLEMS_CHOICE is None:
    problem_indices_choice = list(range(len(dataset["train"])))
else:
    stop_choice = min(len(dataset["train"]), START_INDEX_CHOICE + NUM_PROBLEMS_CHOICE)
    problem_indices_choice = list(range(START_INDEX_CHOICE, stop_choice))

if not problem_indices_choice:
    raise ValueError('No problems selected. Adjust START_INDEX_CHOICE / NUM_PROBLEMS_CHOICE.')

print(f"Evaluating execution choice on {len(problem_indices_choice)} problems with {NUM_RUNS_PER_PROBLEM} run(s) each.")
print(f"Reasoning effort: {REASONING_EFFORT_CHOICE}, temperature: {TEMPERATURE_CHOICE}, top_p: {TOP_P_CHOICE}")

execution_choice_counts = {
    'preference': {'original': 0, 'mutated': 0, 'total': 0},
    'OC': {'correct': 0, 'total': 0, 'reversion_correct': 0, 'reversion_total': 0},
    'MC': {'correct': 0, 'total': 0, 'reversion_correct': 0, 'reversion_total': 0},
    'invalid_runs': 0,
}

execution_choice_results = []
execution_choice_latencies = []
reversion_skip_count_choice = 0

orderings = [True, False]  # True => original program first, False => mutated first
selected_orderings = orderings[:NUM_RUNS_PER_PROBLEM]

for idx in tqdm(problem_indices_choice, desc='Execution choice'):
    sample = dataset['train'][idx]
    original_output = sample['output']
    mutated_output = sample.get('mutated_output')

    include_reversion = True
    if SKIP_BOOLEAN_FOR_REVERSION_CHOICE and (
        is_boolean_output(original_output) or is_boolean_output(mutated_output)
    ):
        include_reversion = False
        reversion_skip_count_choice += 1

    base_seed = SEED_CHOICE + idx * 1000

    for run_offset, original_first in enumerate(selected_orderings):
        prompt, mapping = build_execution_choice_prompt(sample, original_first=original_first)
        generation = generate_choice_response(prompt, base_seed + run_offset)
        execution_choice_latencies.append(generation['latency_s'])

        run_record = {
            'problem_index': int(idx),
            'problem_id': sample['id'],
            'function_name': sample['function_name'],
            'run_index': run_offset,
            'original_first': original_first,
            'response': generation['response'],
            'latency_s': generation['latency_s'],
            'include_reversion': include_reversion,
            'chosen_program_letter': None,
            'chosen_program_type': None,
            'prediction': None,
            'correct_for_chosen_program': None,
            'reversion_for_other_program': None,
            'error': None,
        }

        try:
            parsed = parse_execution_choice_response(generation['response'])
        except ValueError as exc:
            run_record['error'] = str(exc)
            execution_choice_counts['invalid_runs'] += 1
            execution_choice_results.append(run_record)
            continue

        chosen_letter = str(parsed.get('chosen_program', '')).strip().upper()
        assertion_text = parsed.get('assertion', '')
        chosen_type = mapping.get(chosen_letter)

        if chosen_type not in {'original', 'mutated'}:
            run_record['error'] = f"Unrecognized chosen program letter: {chosen_letter}"
            execution_choice_counts['invalid_runs'] += 1
            execution_choice_results.append(run_record)
            continue

        predicted_output = extract_output_from_assertion(assertion_text)
        chosen_output = original_output if chosen_type == 'original' else mutated_output
        other_output = mutated_output if chosen_type == 'original' else original_output

        is_correct, correctness_error = check_predicted_output(predicted_output, chosen_output)
        if include_reversion:
            is_reversion, reversion_error = check_predicted_output(predicted_output, other_output)
        else:
            is_reversion, reversion_error = None, None

        execution_choice_counts['preference']['total'] += 1
        if chosen_type == 'original':
            execution_choice_counts['preference']['original'] += 1
            bucket = execution_choice_counts['OC']
        else:
            execution_choice_counts['preference']['mutated'] += 1
            bucket = execution_choice_counts['MC']

        bucket['total'] += 1
        if is_correct:
            bucket['correct'] += 1
        if include_reversion:
            bucket['reversion_total'] += 1
            if is_reversion:
                bucket['reversion_correct'] += 1

        run_record.update({
            'chosen_program_letter': chosen_letter,
            'chosen_program_type': chosen_type,
            'assertion': assertion_text,
            'prediction': predicted_output,
            'correct_for_chosen_program': bool(is_correct),
            'reversion_for_other_program': bool(is_reversion) if include_reversion else None,
            'correctness_error': correctness_error,
            'reversion_error': reversion_error,
        })

        execution_choice_results.append(run_record)


def _safe_ratio(numerator: int, denominator: int) -> Optional[float]:
    return None if denominator == 0 else numerator / denominator


preference_total = execution_choice_counts['preference']['total']
preference_original_rate = _safe_ratio(
    execution_choice_counts['preference']['original'], preference_total
)
preference_mutated_rate = _safe_ratio(
    execution_choice_counts['preference']['mutated'], preference_total
)

oc_correct_rate = _safe_ratio(
    execution_choice_counts['OC']['correct'], execution_choice_counts['OC']['total']
)
or_reversion_rate = _safe_ratio(
    execution_choice_counts['OC']['reversion_correct'], execution_choice_counts['OC']['reversion_total']
)
mc_correct_rate = _safe_ratio(
    execution_choice_counts['MC']['correct'], execution_choice_counts['MC']['total']
)
mr_reversion_rate = _safe_ratio(
    execution_choice_counts['MC']['reversion_correct'], execution_choice_counts['MC']['reversion_total']
)

execution_choice_summary = {
    'dataset': 'LeetCode',
    'problems_evaluated': len(problem_indices_choice),
    'runs_per_problem': NUM_RUNS_PER_PROBLEM,
    'preference_original': preference_original_rate,
    'preference_mutated': preference_mutated_rate,
    'oc_correct': oc_correct_rate,
    'or_reversion': or_reversion_rate,
    'mc_correct': mc_correct_rate,
    'mr_reversion': mr_reversion_rate,
    'avg_latency_s': (
        sum(execution_choice_latencies) / len(execution_choice_latencies)
        if execution_choice_latencies else None
    ),
    'invalid_runs': execution_choice_counts['invalid_runs'],
    'reversion_skipped_problems': (
        reversion_skip_count_choice if SKIP_BOOLEAN_FOR_REVERSION_CHOICE else 0
    ),
}
choice_metric_pass = {
    'OC': execution_choice_summary['oc_correct'],
    'OR': execution_choice_summary['or_reversion'],
    'MC': execution_choice_summary['mc_correct'],
    'MR': execution_choice_summary['mr_reversion'],
}
choice_metric_counts = {
    'OC': (execution_choice_counts['OC']['correct'], execution_choice_counts['OC']['total']),
    'OR': (execution_choice_counts['OC']['reversion_correct'], execution_choice_counts['OC']['reversion_total']),
    'MC': (execution_choice_counts['MC']['correct'], execution_choice_counts['MC']['total']),
    'MR': (execution_choice_counts['MC']['reversion_correct'], execution_choice_counts['MC']['reversion_total']),
}


execution_choice_table = pd.DataFrame([{
    'Dataset': 'LeetCode',
    'Problems Evaluated': execution_choice_summary['problems_evaluated'],
    'Runs per Problem': execution_choice_summary['runs_per_problem'],
    'Preference (Original)': execution_choice_summary['preference_original'],
    'OC Correct': execution_choice_summary['oc_correct'],
    'OR Reversion': execution_choice_summary['or_reversion'],
    'MC Correct': execution_choice_summary['mc_correct'],
    'MR Reversion': execution_choice_summary['mr_reversion'],
}])

choice_formatters = {
    'Preference (Original)': lambda v, _pd=pd: 'N/A' if _pd.isna(v) else f"{v:.2%}",
    'OC Correct': lambda v, _pd=pd: 'N/A' if _pd.isna(v) else f"{v:.2%}",
    'OR Reversion': lambda v, _pd=pd: 'N/A' if _pd.isna(v) else f"{v:.2%}",
    'MC Correct': lambda v, _pd=pd: 'N/A' if _pd.isna(v) else f"{v:.2%}",
    'MR Reversion': lambda v, _pd=pd: 'N/A' if _pd.isna(v) else f"{v:.2%}",
}

display(execution_choice_table.style.format(choice_formatters))

print('Preference counts:')
print(
    f"  Original: {execution_choice_counts['preference']['original']} / {preference_total}"
)
print(
    f"  Mutated: {execution_choice_counts['preference']['mutated']} / {preference_total}"
)
print(f"Invalid runs (no usable JSON response): {execution_choice_counts['invalid_runs']}")
if execution_choice_summary['avg_latency_s'] is not None:
    print(f"Average latency per run: {execution_choice_summary['avg_latency_s']:.2f}s")
else:
    print('Average latency per run: N/A')
if SKIP_BOOLEAN_FOR_REVERSION_CHOICE:
    print(
        f"Problems skipped for reversion metrics (boolean outputs): {reversion_skip_count_choice}"
    )


Evaluating execution choice on 20 problems with 2 run(s) each.
Reasoning effort: low, temperature: 0.6, top_p: 0.95


Execution choice:   0%|          | 0/20 [00:00<?, ?it/s]

Prompt:  You are given two Python programs below and an assertion containing an input to a function. First, choose either program, whichever one you are more confident in reasoning about. Then, replace the ?? in the assertion with a literal (no unsimplified expressions, no function calls) representing the function's return value for the given input on your chosen program. Execute the program exactly as written, even if it is incorrect or incomplete. For your final answer, output the letter of your chosen program (A or B) and the full assertion in the following json format:

{{
  "chosen_program": "A or B",
  "assertion": "full_assertion"
}}

[PROGRAM_A]
def maxLength(nums: List[int]) -> int:
        n = len(nums)
        ans = 0
        for l in range(n):
            a = 1
            b = 1
            c = 0
            for r in range(l, n):
                a *= nums[r]
                b = lcm(b, nums[r])
                c = gcd(c, nums[r])
                if a == b * c:
              

Preference counts:
  Original: 34 / 40
  Mutated: 6 / 40
Invalid runs (no usable JSON response): 0
Average latency per run: 51.78s
Problems skipped for reversion metrics (boolean outputs): 0


## Step 9: Compare Reasoning Efforts (Optional)

Evaluate a small subset with different reasoning effort settings.

In [ ]:
# Compare low vs medium vs high reasoning
reasoning_levels = ["low", "medium", "high"]
comparison_results = {}

NUM_COMPARISON_SAMPLES = 5  # Use small number for comparison
MAX_NEW_TOKENS = 1000  # Allow enough tokens for reasoning + answer

for reasoning_effort in reasoning_levels:
    print(f"\nTesting reasoning_effort={reasoning_effort}...")
    correct = 0
    total_latency = 0

    for idx in range(NUM_COMPARISON_SAMPLES):
        sample = dataset['train'][idx]

        prompt = build_execution_prediction_prompt(sample)
        messages = [
            {"role": "user", "content": prompt}
        ]

        inputs = tokenizer.apply_chat_template(
            messages,
            add_generation_prompt=True,
            return_tensors="pt",
            return_dict=True,
            reasoning_effort=reasoning_effort,
        ).to(model.device)

        start_time = time.time()
        generated = model.generate(**inputs, max_new_tokens=MAX_NEW_TOKENS)
        latency = time.time() - start_time
        total_latency += latency

        response = tokenizer.decode(
            generated[0][inputs["input_ids"].shape[-1]:],
            skip_special_tokens=True
        )

        predicted_output = extract_answer_from_response(response)
        is_correct, _ = check_predicted_output(
            predicted_output, sample['output']
        )

        if is_correct:
            correct += 1

    comparison_results[reasoning_effort] = {
        "correct": correct,
        "pass@1": correct / NUM_COMPARISON_SAMPLES,
        "avg_latency": total_latency / NUM_COMPARISON_SAMPLES
    }

# Print comparison
print("\n" + "="*60)
print("REASONING EFFORT COMPARISON")
print("="*60)
print(f"{'Reasoning':<12} {'pass@1':<10} {'Avg Latency':<15}")
print("-" * 60)
for level, stats in comparison_results.items():
    print(f"{level:<12} {stats['pass@1']*100:>6.1f}%   {stats['avg_latency']:>10.2f}s")
print("="*60)

## Next Steps

1. **Increase NUM_SAMPLES** to evaluate on more problems
2. **Try different reasoning_effort** levels
3. **Compare with other models** (DeepSeek-R1, GPT-4o, etc.)
4. **Analyze error patterns** to understand model weaknesses

---

**Note**: To run on full dataset (347 samples), expect ~1-2 hours on free Colab.